## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [96]:
import pandas as pd

df = pd.read_csv('./Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [97]:
# View the column names in the data
column_header_data = df.columns
print(column_header_data)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [98]:
# Use the describe function to gather some basic statistics
basic_stats = df.describe(include='all')
print(basic_stats)

              first     last        job       phone                     email  \
count         54639    54639      54639       54639                     54639   
unique          360      477        498         999                       999   
top     Christopher  Johnson  Osteopath  5016915923  jessica.reyes@grimes.net   
freq           1090     1164        631         220                       220   
mean            NaN      NaN        NaN         NaN                       NaN   
std             NaN      NaN        NaN         NaN                       NaN   
min             NaN      NaN        NaN         NaN                       NaN   
25%             NaN      NaN        NaN         NaN                       NaN   
50%             NaN      NaN        NaN         NaN                       NaN   
75%             NaN      NaN        NaN         NaN                       NaN   
max             NaN      NaN        NaN         NaN                       NaN   

           client_id      o

In [99]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# shape_df = df.shape
# print(shape_df)

# size_df = df.size
# print(size_df)

# type_df = df.dtypes
# print(type_df)

info_df = df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [100]:
# What three item categories had the most entries?
# sort_data_by_cat = df["category"].sort_values()
# reset_index_sorted_data = sort_data_by_cat.reset_index(drop=True)
value_counts_df = df["category"].value_counts()
# print(sort_data_by_cat)
# print(reset_index_sorted_data)
top_3_counts_df = value_counts.head(3)
# print(value_counts_df)
print(top_3_counts_df)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [112]:
# For the category with the most entries,
# which subcategory had the most entries?
sub_value_counts_df = df[["category", "subcategory"]].value_counts()
# top_sub_counts_df = sub_value_counts_df.head(1)
sub_value_counts_df.head(1)
# print(top_sub_counts_df)

category     subcategory      
consumables  bathroom supplies    6424
Name: count, dtype: int64

In [111]:
# Which five clients had the most entries in the data?
top_5_clients_df = df[["first", "last", "client_id"]].value_counts()
top_5_clients_df.head()

first      last     client_id
Jessica    Reyes    33615        220
Angela     Everett  66037        211
Bryan      Myers    46820        209
Kendra     Garrett  24741        207
Alexandra  Young    38378        207
Name: count, dtype: int64

In [129]:
# Store the client ids of those top 5 clients in a list.
top_clients_id_df = df["client_id"].value_counts()
top_5_clients_df = top_clients_id_df.head(5)
top_5_clients_df.head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [141]:
# How many total units (the qty column) did the
# client with the most entries order order?
client_qty_ordered_df = df.groupby("client_id")["qty"].sum()
is_client_id = 33615
is_client_qty = 220
client_known_data = client_qty_ordered_df.loc[is_client_id]
print("Client " + str(is_client_id) + " has ordered " + str(is_client_qty) + " times, for a total of " + str(client_known_data) + " items.")

Client 33615 has ordered 220 times, for a total of 64313 items.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [ ]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty



In [ ]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.



In [ ]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%



In [ ]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).



In [ ]:
# Create a column for the profit of
# each line using line cost and line price



## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
# Check your work using the totals above


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [ ]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.



In [ ]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



In [ ]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
